File: lm/ngram/smoothing/kneser_ney.py
import lm.ngram.model as ngram


class KneserNey(ngram.NGramLM):

    def __init__(self, n, discount):
        super().__init__(n)
        self._discount = discount

    def prob(self, word, word_seq):
        trimmed_word_seq = word_seq[-self._n+1:]

        if len(word_seq) == 0:
            return self._trie.count_preceding([word]) / self._trie.total_seqs_of_len(2)
        else:
            max_numerator = max(self._trie.count(trimmed_word_seq + [word]) - self._discount, 0)
            sum_following = float(self._trie.sum_following(trimmed_word_seq))
            num_following = self._trie.count_following(trimmed_word_seq)

            if sum_following == 0:
                return 0
            else:
                return (max_numerator + self._discount * num_following * self._prob_kn(word, trimmed_word_seq[1:])) /\
                       sum_following

    def _prob_kn(self, word, word_seq):
        l = len(word_seq)
        assert l >= 0
        assert l < self._n

        if l == 0:
            return self._trie.count_preceding([word]) / self._trie.total_seqs_of_len(2)

        num_preceding = self._trie.count_preceding(word_seq)
        num_preceding_and_following = float(self._trie.count_preceding_and_following(word_seq))
        num_following = self._trie.count_following(word_seq)

        max_numerator = max(num_preceding - self._discount, 0)

        if num_preceding_and_following == 0:
            return 0
        else:
            return (max_numerator +
                self._discount * num_following * self._prob_kn(word, word_seq[1:])) / num_preceding_and_following

File: lm/ngram/smoothing/kneser_ney.cc
#include "kneser_ney.h"

KneserNey::KneserNey(std::string file_name, int n, int min_frequency) : AbsoluteDiscounting(n, min_frequency) {
    ProcessFile(file_name);
}

tensorflow::Source::lm::ngram::NGramProto *KneserNey::ToProto() {
    tensorflow::Source::lm::ngram::NGramProto *ngram_proto = NGram::ToProto();
    ngram_proto->set_smoothing(tensorflow::Source::lm::ngram::Smoothing::KNESER_NEY);
    ngram_proto->add_discount(discount);
    return ngram_proto;
}

void KneserNey::PopulateProbTrie(CountTrie *countTrie, CountTrie::Node *node, int depth, std::list<size_t> seq) {
    if (depth > 0) {

        // Calculate pseudo_prob.
        double pseudo_prob = 0;
        double pseudo_prob_numerator = 0;
        double pseudo_prob_denominator = 0;
        if (depth == 1) {
            pseudo_prob_numerator = countTrie->CountPreceding(seq);
            pseudo_prob_denominator = countTrie->CountPrecedingAndFollowing(std::list<size_t>({}));
        } else if (depth == n) {
            int count = countTrie->Count(seq);
            pseudo_prob_numerator = std::max(count - Discount(depth, count), 0.0);
            size_t last_word_index = seq.back();
            seq.pop_back();
            pseudo_prob_denominator = countTrie->SumFollowing(seq);
            seq.push_back(last_word_index);
        } else {
            pseudo_prob_numerator = std::max(countTrie->CountPreceding(seq) - Discount(depth, countTrie->Count(seq)), 0.0);
            size_t last_word_index = seq.back();
            seq.pop_back();
            pseudo_prob_denominator = countTrie->CountPrecedingAndFollowing(seq);
            seq.push_back(last_word_index);
        }
        if (pseudo_prob_denominator > 0) {
            pseudo_prob = pseudo_prob_numerator / pseudo_prob_denominator;
        }

        // Calculate backoff.
        double backoff = 1;
        double backoff_denominator = 0;
        if (depth == n - 1) {
            backoff_denominator = countTrie->SumFollowing(seq);
        } else {
            backoff_denominator = countTrie->CountPrecedingAndFollowing(seq);
        }
        if (backoff_denominator > 0) {
            backoff = BackoffNumerator(countTrie, depth, seq) / backoff_denominator;
        }

        prob_trie->Insert(seq, pseudo_prob, backoff);
    }
    for (std::unordered_map<size_t, CountTrie::Node*>::iterator it = node->children.begin(); it != node->children.end(); ++it) {
        seq.push_back(it->first);
        PopulateProbTrie(countTrie, it->second, depth + 1, seq);
        seq.pop_back();
    }
}

double KneserNey::Discount(int n, int count) {
    return discount;
}

double KneserNey::BackoffNumerator(CountTrie *count_trie, int depth, std::list<size_t> seq) {
    return discount * count_trie->CountFollowing(seq);
}
